test: migrate weak_dep_features, workspaces and yank to snapbox

eth3lbert opened the pull request:
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/14039.

Migrate following to snapbox:

- `tests/testsuite/weak_dep_features.rs`
- `tests/testsuite/workspaces.rs`
- `tests/testsuite/yank.rs`

epage commented on tests/testsuite/workspaces.rs:
```diff
@@ -430,14 +424,13 @@ fn invalid_parent_pointer() {
 Caused by:
-  [..]
-",
-        )
+  [..] (os error [..])
```
Why was this glob made more specific?

eth3lbert replied:
Originally, I attempted a more specific approach without using any globs. However, this resulted in different error messages on different platforms https://github.com/rust-lang/cargo/actions/runs/9587495931/job/26437670924#step:11:4084.

epage replied:
We have an auto-redaction for `[BROKEN_PIPE]` that matches the text for each platform's broken pipe message.  We could do something similar here.  As this is a common error message, auto-redacting it would likely help a lot!

eth3lbert replied:
Sounds good!

One question: How should we handle situations where the same error occurs on Darwin/Linux but potentially results in different errors on Windows?

e.g.
https://github.com/rust-lang/cargo/actions/runs/9587495931/job/26437670924#step:11:4065
```
   7    7 | Caused by:
   8      -   No such file or directory (os error 2)
        8 +   The system cannot find the file specified. (os error 2)
```
https://github.com/rust-lang/cargo/actions/runs/9587495931/job/26437670924#step:11:4084
```
   7    7 | Caused by:
   8      -   No such file or directory (os error 2)
        8 +   The system cannot find the path specified. (os error 3)
 ```

weihanglo replied:
There is another one https://github.com/rust-lang/cargo/pull/14068/files#r1639170882

epage replied:
See https://github.com/rust-lang/cargo/blob/ce90393a11758106040628f77568337b1a5b3a0e/crates/cargo-test-support/src/compare.rs#L202-L203

Define the redaction multiple times, once for each unique text string.  This should be in `MIN_LITERAL_REDACTIONS`

epage commented on tests/testsuite/workspaces.rs:
```diff
@@ -1251,7 +1230,14 @@ fn rebuild_please() {
         .cwd("bin")
         .with_status(101)
-        .with_stderr_contains("[..]assertion[..]")
+        .with_stderr_data(str![[r#"
+...
+assertion `left == right` failed
+  left: 1
+ right: 0
+[NOTE] run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+...
+"#]])
```
We try to minimize rustc/std output that we match because we support multiple versions of rustc and changing rustc's behavior can be more difficult when cargo tests are dependent on it.  We should probably match on as little as possible.

For some cases, we don't even think about this.  I'm only thinking about this one because the output recently changed.

eth3lbert replied:
Ok, I'll update to use the original matching pattern.

eth3lbert commented:
Rebased and feedback addressed!

Additionally, I've replaced an empty `str![]` with `""`.
```diff
diff --git a/tests/testsuite/weak_dep_features.rs b/tests/testsuite/weak_dep_features.rs
index f5d2e18e5..da281bfe3 100644
--- a/tests/testsuite/weak_dep_features.rs
+++ b/tests/testsuite/weak_dep_features.rs
@@ -517,7 +517,7 @@ foo v0.1.0 ([ROOT]/foo)
     // This is a little strange in that it produces no output.
     // Maybe `cargo tree` should print a note about why?
     p.cargo("tree -e features -i bar --features bar?/feat")
-        .with_stdout_data(str![])
+        .with_stdout_data("")
         .run();
 
     p.cargo("tree -e features -i bar --features bar?/feat,bar")
```

Thanks for the review! :)

epage commented:
@bors r+

bors commented:
:pushpin: Commit e3ea0aad3024fb98579168ba413ccb51d42ac829 has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage e3ea0aad3024fb98579168ba413ccb51d42ac829 -->
<!-- homu: {"type":"Approved","sha":"e3ea0aad3024fb98579168ba413ccb51d42ac829","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit e3ea0aad3024fb98579168ba413ccb51d42ac829 with merge bc89bffa5987d4af8f71011c7557119b39e44a65...
<!-- homu: {"type":"BuildStarted","head_sha":"e3ea0aad3024fb98579168ba413ccb51d42ac829","merge_sha":"bc89bffa5987d4af8f71011c7557119b39e44a65"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/9621434386/job/26541841600)
Approved by: epage
Pushing bc89bffa5987d4af8f71011c7557119b39e44a65 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/9621434386/job/26541841600"},"merge_sha":"bc89bffa5987d4af8f71011c7557119b39e44a65"} -->
